### PR TITLE
Refactor covariance matrix calculations:

### PR DESCRIPTION
**Pull request by mmarcinkiewicz**
Refactor covariance matrix calculations:
- use only the most current activation
- change default sampling rate from B
- flatten B, H, W for Conv layers instead of median

Fix calculation of number of eigval by argmax.

Change CIFAR10 training schedule for faster convergence (the loss seemed not to drop almost at all during training example_deep.py):
- normalization function
- batch size
- learning rate
- h2 sizes (due to different saturation calculations)

### CODE REVIEW

**Comment by justinshenk** [comment]
Looks good, thanks Michal!

**Comment by justinshenk** [comment]
Hey Michal,

If you would like to make edits or editions to the paper, please feel free
to make them directly to the LaTeX document:
https://v1.overleaf.com/16149280fzmmrctbpddg

Let me know what when you make edits so that I can add you as coauthor for
an upcoming submission.

Regards,
Justin

**Comment by mmarcinkiewicz** [comment]
Hey Justin!

I am making some corrections right now. Preferably I'd put them as a
comment so you can agree or not. Unless you can revert it easily to the
original version?

Best,
MM

czw., 6 gru 2018 o 15:52 Justin Shenk <notifications@github.com> napisał(a):

> Hey Michal,
>
> If you would like to make edits or editions to the paper, please feel free
> to make them directly to the LaTeX document:
> https://v1.overleaf.com/16149280fzmmrctbpddg
>
> Let me know what when you make edits so that I can add you as coauthor for
> an upcoming submission.
>
> Regards,
> Justin
>
> On Fri, Nov 30, 2018 at 6:07 AM Michał Marcinkiewicz <

**Comment by justinshenk** [comment]
Hi Michal,
Thanks for the heads up. You can make changes directly, just please comment
them so I know what is changed.

Justin

**Comment by justinshenk** [comment]
Michal,

I just realised comments aren't necessary, actually, since I can do the git
diff with the backup.

Thanks,
Justin

**Comment by mmarcinkiewicz** [comment]
Justin,

ok, cool. I've just made an account, so now maybe the changes are better
traceable.

Best,
MM

czw., 6 gru 2018 o 18:07 Justin Shenk <notifications@github.com> napisał(a):

> Michal,
>
> I just realised comments aren't necessary, actually, since I can do the git
> diff with the backup.
>
> Thanks,
> Justin
>
> On Thu, Dec 6, 2018 at 6:05 PM Justin Shenk <shenk.justin@gmail.com>
> wrote:
>
> > Hi Michal,
> > Thanks for the heads up. You can make changes directly, just please
> > comment them so I know what is changed.
> >
> > Justin
> >
> > On Thu, Dec 6, 2018 at 6:02 PM Michał Marcinkiewicz <

**Comment by justinshenk** [comment]
Alright. Let me know if you have any questions about the paper or results.
The paper was a bit rushed for a deadline (to AI Stats, but was rejected)
so lots of room for improvement at this stage.

Justin

**Comment by mmarcinkiewicz** [comment]
Yes, I will.
I started to change things, then I'll realise that something in my
understanding is off. I'll have to think about it before I can proceed. I
reverted the file to the original state (or almost original).
I'll let you know if I need a hand!

Cheers,
MM

czw., 6 gru 2018 o 18:23 Justin Shenk <notifications@github.com> napisał(a):

> Alright. Let me know if you have any questions about the paper or results.
> The paper was a bit rushed for a deadline (to AI Stats, but was rejected)
> so lots of room for improvement at this stage.
>
> Justin
>
> On Thu, Dec 6, 2018 at 6:10 PM Michał Marcinkiewicz <

**Comment by justinshenk** [comment]
Thanks for the update. If something is not clear, let me know.

**Comment by mmarcinkiewicz** [comment]
Hey Justin,

yes, I could use a clarification, because I thought I had it figured out,
but I was wrong:
1. I don't actually get the point of making statistics on samples from
previous time steps. What are those time steps - previous epochs or batches?
I thought the general idea is to check whether the layer is capable of
representing the data efficiently (in terms of number of principal
components required to explain the variance). And by data I understand the
current batch fed into the layer, is it correct? Then I don't see how it
should depend on the previous batch (sorry, I am having problems tracking
down what layer_history property is and when it is being set).

2. If I'm wrong please ignore that point and the one below.
But if I am right - let's have a look at the current activations for each
type of layer, starting from Fully-Connected. As far as I know, a FC layer
takes n dimensional input and casts it to d dimensional output in some
particular way. Now, if we used batch size = 1, we are stuck with a vector
of d pre-activations. I don't actually see what the covariance of this
vector would represent and what would an eigendecomposition of such matrix
be. We of course could (as we do now) take a larger batch and say that it
represents the whole dataset (which let's say is a valid approximation) and
see if we can do covariance between representations of different
observations in a batch. But since different observation in a batch can
belong to different classes of objects, does it make sense to calculate
covariance between them?

3. If it comes to CNN I think I have a better understanding, please verify.
Because now we can change our view on the dataset. Instead of having one
dataset with m observations, we can assume that each filter f in a layer
extracts information whether a given feature is present at a given h,w of
the particular image. Now the image becomes "a whole field (H*W) of
observations", and each observation has f features, we can check if
potentially we could get around with less features (filters) to represent
the same amount of variance to represent data at every h and w. And that
makes sense, even more - because we don't actually care about the batch
size. We treat each individual image as a dataset on its own.

I'd appreciate your feedback about it, because I got lost somewhere on the
way.

Cheers,
Michal

czw., 6 gru 2018 o 20:03 Justin Shenk <notifications@github.com> napisał(a):

> Thanks for the update. If something is not clear, let me know.
>
> On Thu, Dec 6, 2018 at 7:22 PM Michał Marcinkiewicz <
